File: packages/sdk-java/src/main/java/org/melaya/Melaya.java
package org.melaya;

/**
 * The Melaya client — entry point for all SDK operations.
 *
 * <h2>Namespaced API (recommended)</h2>
 * <pre>{@code
 * Melaya melaya = new Melaya("mk_yourkey");
 *
 * // ── trading ──────────────────────────────────────────────────────────────
 * JsonNode ticker   = melaya.trading().market().ticker("binance", "BTC/USDT", "spot");
 * melaya.trading().trade().createOrder(...);                   // real funds
 * melaya.trading().backtest().start(body);
 * try (MelayaStream s = melaya.trading().stream().ticker("binance","BTC/USDT","spot")) { ... }
 *
 * // ── agents ───────────────────────────────────────────────────────────────
 * melaya.agents().pipelines().traces("run-123");
 * JsonNode pending = melaya.agents().hitl().pending();
 * melaya.agents().hitl().approve(pending.get(0).get("requestId").asText(), null);
 * melaya.agents().evals().summary();
 *
 * // ── platform ─────────────────────────────────────────────────────────────
 * JsonNode projects = melaya.platform().projects().list();
 * melaya.platform().credentials().set("openai", Map.of("value", "sk-..."));
 * melaya.platform().billing().subscription();
 * melaya.platform().events().onRunUpdate("run-123", frame ->
 *     System.out.println(frame.get("event_type").asText()));
 * }</pre>
 *
 * <h2>Flat API (legacy, also valid)</h2>
 * <pre>{@code
 * // Equivalent flat accessors are preserved for backward compatibility:
 * melaya.market().ticker("binance", "BTC/USDT", "spot");
 * melaya.projects().list();
 * melaya.hitl().pending();
 * melaya.events().onRunUpdate("run-123", frame -> System.out.println(frame));
 * }</pre>
 *
 * <p><strong>REST base URL:</strong> {@code https://api.melaya.org}<br>
 * <strong>WS base URL:</strong> {@code wss://wss.melaya.org}
 *
 * <p>Every REST call injects the API key as:
 * <ul>
 *   <li>{@code Authorization: Bearer mk_...} header</li>
 * </ul>
 *
 * <p><strong>TLS:</strong> certificate and hostname verification always use the
 * JVM trust store; verification is never disabled.
 *
 * <p><strong>Retries:</strong> only idempotent GET requests are retried — up to
 * 2 extra attempts on network errors, 429, and 5xx responses, with exponential
 * back-off (500 ms → 1 s, plus up to 200 ms of random jitter), respecting the
 * {@code Retry-After} header when present. Non-GET methods are never retried.
 *
 * <p><strong>Security:</strong> API keys and JWT tokens are never logged.
 * Secrets are only carried in {@code Authorization} headers, never echoed in
 * exception messages.
 */
public class Melaya {

    /** REST base URL. */
    public static final String DEFAULT_BASE_URL = "https://api.melaya.org";
    /** WebSocket base URL. */
    public static final String DEFAULT_WS_URL   = "wss://wss.melaya.org";

    // ── Namespace accessors (primary API) ─────────────────────────────────────
    private final TradingNamespace  trading;
    private final AgentsNamespace   agents;
    private final PlatformNamespace platform;

    // ── Trading plane (flat — kept for backward compatibility) ────────────────
    private final MarketAPI     market;
    private final AccountAPI    account;
    private final SimAPI        sim;
    private final TradeAPI      trade;
    private final StrategiesAPI strategies;
    private final BacktestAPI   backtest;
    private final StreamAPI     stream;

    // ── Platform / agents plane (flat — kept for backward compatibility) ──────
    private final AuthAPI        auth;
    private final MfaAPI         mfa;
    private final AccountsAPI    accounts;
    private final BillingAPI     billing;
    private final RunnerAPI      runner;
    private final ProjectsAPI    projects;
    private final PipelinesAPI   pipelines;
    private final HitlAPI        hitl;
    private final CredentialsAPI credentials;
    private final ConnectorsAPI  connectors;
    private final OverviewAPI    overview;
    private final PhoneAPI       phone;
    private final TeamAPI        team;
    private final TemplatesAPI   templates;
    private final AssistantAPI   assistant;
    private final EvalsAPI       evals;
    private final BugsAPI        bugs;
    private final MelayaEvents   events;

    /**
     * Create a Melaya client with the given API key.
     *
     * @param apiKey your Melaya API key (must start with {@code mk_})
     * @throws IllegalArgumentException if the key is blank or not prefixed {@code mk_}
     */
    public Melaya(String apiKey) {
        this(apiKey, DEFAULT_BASE_URL, DEFAULT_WS_URL);
    }

    /**
     * Create a Melaya client with custom base URLs (useful for testing or staging).
     *
     * @param apiKey  your Melaya API key (must start with {@code mk_})
     * @param baseUrl REST base URL override
     * @param wsUrl   WebSocket base URL override
     */
    public Melaya(String apiKey, String baseUrl, String wsUrl) {
        if (apiKey == null || apiKey.isBlank()) {
            throw new IllegalArgumentException(
                    "Melaya: apiKey is required (create one at melaya.org → Settings → API Keys).");
        }
        if (!apiKey.startsWith("mk_")) {
            throw new IllegalArgumentException("Melaya: API keys must be prefixed 'mk_'.");
        }
        HttpClient http = new HttpClient(apiKey, baseUrl);
        this.market      = new MarketAPI(http);
        this.account     = new AccountAPI(http);
        this.sim         = new SimAPI(http);
        this.trade       = new TradeAPI(http);
        this.strategies  = new StrategiesAPI(http);
        this.backtest    = new BacktestAPI(http);
        this.stream      = new StreamAPI(apiKey, wsUrl, http);
        this.auth        = new AuthAPI(http);
        this.mfa         = new MfaAPI(http);
        this.accounts    = new AccountsAPI(http);
        this.billing     = new BillingAPI(http);
        this.runner      = new RunnerAPI(http);
        this.projects    = new ProjectsAPI(http);
        this.pipelines   = new PipelinesAPI(http);
        this.hitl        = new HitlAPI(http);
        this.credentials = new CredentialsAPI(http);
        this.connectors  = new ConnectorsAPI(http);
        this.overview    = new OverviewAPI(http);
        this.phone       = new PhoneAPI(http);
        this.team        = new TeamAPI(http);
        this.templates   = new TemplatesAPI(http);
        this.assistant   = new AssistantAPI(http);
        this.evals       = new EvalsAPI(http);
        this.bugs        = new BugsAPI(http);
        this.events      = new MelayaEvents(apiKey, baseUrl);
        // ── Namespace wrappers (share the already-constructed module instances) ──
        this.trading  = new TradingNamespace(
                this.market, this.account, this.sim,
                this.strategies, this.backtest, this.trade, this.stream);
        this.agents   = new AgentsNamespace(
                this.pipelines, this.hitl, this.assistant, this.phone, this.evals);
        this.platform = new PlatformNamespace(
                this.projects, this.credentials, this.connectors, this.billing,
                this.team, this.templates, this.overview, this.runner,
                this.auth, this.mfa, this.accounts, this.bugs, this.events);
    }

    /** Package-private constructor for injecting a pre-built HttpClient (e.g. E2E). */
    Melaya(HttpClient http, String wsUrl) {
        String apiKey = http.getApiKey();
        String baseUrl = http.getBaseUrl();
        this.market      = new MarketAPI(http);
        this.account     = new AccountAPI(http);
        this.sim         = new SimAPI(http);
        this.trade       = new TradeAPI(http);
        this.strategies  = new StrategiesAPI(http);
        this.backtest    = new BacktestAPI(http);
        this.stream      = new StreamAPI(apiKey, wsUrl, http);
        this.auth        = new AuthAPI(http);
        this.mfa         = new MfaAPI(http);
        this.accounts    = new AccountsAPI(http);
        this.billing     = new BillingAPI(http);
        this.runner      = new RunnerAPI(http);
        this.projects    = new ProjectsAPI(http);
        this.pipelines   = new PipelinesAPI(http);
        this.hitl        = new HitlAPI(http);
        this.credentials = new CredentialsAPI(http);
        this.connectors  = new ConnectorsAPI(http);
        this.overview    = new OverviewAPI(http);
        this.phone       = new PhoneAPI(http);
        this.team        = new TeamAPI(http);
        this.templates   = new TemplatesAPI(http);
        this.assistant   = new AssistantAPI(http);
        this.evals       = new EvalsAPI(http);
        this.bugs        = new BugsAPI(http);
        this.events      = new MelayaEvents(apiKey, baseUrl);
        // ── Namespace wrappers (share the already-constructed module instances) ──
        this.trading  = new TradingNamespace(
                this.market, this.account, this.sim,
                this.strategies, this.backtest, this.trade, this.stream);
        this.agents   = new AgentsNamespace(
                this.pipelines, this.hitl, this.assistant, this.phone, this.evals);
        this.platform = new PlatformNamespace(
                this.projects, this.credentials, this.connectors, this.billing,
                this.team, this.templates, this.overview, this.runner,
                this.auth, this.mfa, this.accounts, this.bugs, this.events);
    }

    // ── Namespace accessors (primary API) ────────────────────────────────────

    /**
     * Trading namespace — market data, CEX account, sim, strategies, backtest,
     * live trade, and WebSocket streaming.
     *
     * <pre>{@code
     * melaya.trading().market().ticker("binance", "BTC/USDT", "spot");
     * melaya.trading().trade().createOrder(...);
     * melaya.trading().stream().orderbook("binance", "BTC/USDT", "spot", 20);
     * }</pre>
     */
    public TradingNamespace trading() { return trading; }

    /**
     * Agents namespace — pipeline runs, HITL approvals, assistant profile,
     * phone device control, and evaluations.
     *
     * <pre>{@code
     * melaya.agents().pipelines().traces("run-123");
     * melaya.agents().hitl().approve("requestId", null);
     * melaya.agents().evals().summary();
     * }</pre>
     */
    public AgentsNamespace agents() { return agents; }

    /**
     * Platform namespace — projects, credentials, connectors, billing, team,
     * templates, overview, runner, auth, MFA, accounts, bugs, and real-time events.
     *
     * <pre>{@code
     * melaya.platform().projects().list();
     * melaya.platform().credentials().set("openai", Map.of("value", "sk-..."));
     * melaya.platform().events().onRunUpdate("run-123", frame -> System.out.println(frame));
     * }</pre>
     */
    public PlatformNamespace platform() { return platform; }

    // ── Flat accessors (preserved for backward compatibility) ─────────────────

    /** REST market-data + reference endpoints (public + authenticated). */
    public MarketAPI market() { return market; }

    /** Authenticated account reads: connected keys, tier limits, usage. */
    public AccountAPI account() { return account; }

    /** Paper trading (sim broker): virtual balance, positions, and orders. */
    public SimAPI sim() { return sim; }

    /** Live trading (real funds): order placement, positions, balance on a connected venue. */
    public TradeAPI trade() { return trade; }

    /** Launch, control, and inspect trading strategies (paper + live). */
    public StrategiesAPI strategies() { return strategies; }

    /** Historical backtests + parameter sweeps on the Rust engine. */
    public BacktestAPI backtest() { return backtest; }

    /** WebSocket streaming endpoints (public market data + private feeds). */
    public StreamAPI stream() { return stream; }

    // ── Platform / agents plane accessors ─────────────────────────────────────

    /**
     * Auth API — session management, login, registration, password reset, token refresh.
     */
    public AuthAPI auth() { return auth; }

    /** MFA API — enroll, confirm, and check TOTP. */
    public MfaAPI mfa() { return mfa; }

    /**
     * Accounts API — profile update, GDPR export, API-key removal, credit balances.
     * Distinct from {@link #account()} which covers trading account reads.
     */
    public AccountsAPI accounts() { return accounts; }

    /** Billing API — subscription status, Stripe checkout/portal, pricing plans. */
    public BillingAPI billing() { return billing; }

    /** Runner token API — mint, list, and revoke {@code mel_run_} tokens. */
    public RunnerAPI runner() { return runner; }

    /** Projects API — create and list agent projects. */
    public ProjectsAPI projects() { return projects; }

    /** Pipelines API — run overview, traces, and cron schedules. */
    public PipelinesAPI pipelines() { return pipelines; }

    /** HITL API — list, approve, and reject pending tool-call approvals. */
    public HitlAPI hitl() { return hitl; }

    /** Credentials API — user-scoped credential storage, OAuth, RAG, models. */
    public CredentialsAPI credentials() { return credentials; }

    /** Connectors API — project-scoped connector credentials. */
    public ConnectorsAPI connectors() { return connectors; }

    /** Overview API — dashboard metrics, cost breakdown, pipeline lists. */
    public OverviewAPI overview() { return overview; }

    /** Phone API — pair and control Android devices via the Melaya APK. */
    public PhoneAPI phone() { return phone; }

    /** Team API — project team membership, roles, and invitations. */
    public TeamAPI team() { return team; }

    /** Templates API — create, share, assign, and manage pipeline templates. */
    public TemplatesAPI templates() { return templates; }

    /** Assistant API — get and set the caller's onboarding/persona profile. */
    public AssistantAPI assistant() { return assistant; }

    /** Evals API — list, compare, and inspect pipeline evaluation results. */
    public EvalsAPI evals() { return evals; }

    /** Bugs API — submit and track bug reports. */
    public BugsAPI bugs() { return bugs; }

    /**
     * Platform real-time events over Socket.IO at {@code /api/v1/events}.
     * Subscribe to run updates, init-phase progress, HITL notifications,
     * and pipeline CRUD events.
     */
    public MelayaEvents events() { return events; }
}

File: packages/sdk-kotlin/src/main/kotlin/org/melaya/Melaya.kt
package org.melaya

import okhttp3.OkHttpClient
import java.util.concurrent.TimeUnit

const val DEFAULT_BASE_URL = "https://api.melaya.org"
const val DEFAULT_WS_URL   = "wss://wss.melaya.org"

/**
 * The Melaya SDK entry point — v0.2.0.
 *
 * Covers the full REST surface across trading, agents, and platform planes,
 * plus real-time events via Socket.IO at `/api/v1/events`.
 *
 * Authentication: pass your `mk_` platform API key. REST requests send it
 * only as an `Authorization: Bearer` header. Public market WebSockets use the
 * server-required query parameter; private WebSockets use short-lived tickets.
 * The API key is never logged.
 *
 * ### Namespaced API (primary)
 *
 * The three top-level namespaces group every module by plane so IDE autocomplete
 * surfaces the structure immediately:
 *
 * ```kotlin
 * val melaya = Melaya(apiKey = System.getenv("MK"))
 *
 * // ── Trading plane ──────────────────────────────────────────────────────
 * val ticker = melaya.trading.market.ticker("binance", "BTC/USDT", "spot")
 * val book   = melaya.trading.stream.orderbook("bybit", "BTC/USDT", "spot")
 * melaya.trading.strategies.list()
 * melaya.trading.backtest.start(…)
 *
 * // ── Agents plane ───────────────────────────────────────────────────────
 * val runs    = melaya.agents.pipelines.recent()
 * val pending = melaya.agents.hitl.pending()
 * melaya.agents.hitl.approve(pending[0].getString("requestId"))
 *
 * // ── Platform plane ─────────────────────────────────────────────────────
 * val projects = melaya.platform.projects.list()
 * melaya.platform.credentials.set("openai", value = "sk-…")
 * melaya.platform.billing.subscription()
 *
 * // ── Real-time events ───────────────────────────────────────────────────
 * val unsub = melaya.platform.events.onRunUpdate("run-123") { e ->
 *     println("${e.optString("event_type")}: ${e.optString("status")}")
 * }
 * unsub()
 * melaya.platform.events.close()
 * ```
 *
 * ### Flat accessors (legacy / convenience)
 *
 * All modules remain directly accessible on the root client for backwards
 * compatibility: `melaya.market`, `melaya.hitl`, `melaya.projects`, etc.
 *
 * TLS certificate and hostname verification are always enabled. Install a
 * private development CA in the JVM trust store when using an intercept proxy.
 */
class Melaya @JvmOverloads constructor(
    apiKey: String,
    baseUrl: String = DEFAULT_BASE_URL,
    wsUrl: String   = DEFAULT_WS_URL,
    /**
     * Per-request HTTP call timeout in milliseconds.
     * Applied to every REST call via a per-call OkHttp client override.
     * Default: 30 000 ms. Set to 0 to disable.
     */
    requestTimeoutMs: Long = 30_000L,
) {
    init {
        require(apiKey.isNotBlank()) {
            "Melaya: apiKey is required (create one at melaya.org → Settings → API Keys)."
        }
        require(apiKey.startsWith("mk_")) {
            "Melaya: API keys must be prefixed 'mk_'."
        }
    }

    private val okHttp: OkHttpClient = buildOkHttpClient()
    private val http = HttpClient(apiKey, baseUrl, okHttp, requestTimeoutMs)

    // ── Trading plane ──────────────────────────────────────────────────────────

    /** REST market-data + reference endpoints (public and authenticated plane). */
    val market = MarketAPI(http)

    /** Authenticated account reads: connected exchange keys, tier limits, usage. */
    val account = AccountAPI(http)

    /** Paper trading (sim broker): virtual balance, positions, and orders. */
    val sim = SimAPI(http)

    /** Live trading (real funds): order placement, positions, balance on a connected venue. */
    val trade = TradeAPI(http)

    /** Launch, control, and inspect trading strategies (paper + live). */
    val strategies = StrategiesAPI(http)

    /** Historical backtests + parameter sweeps (including optimization runs) on the Rust engine. */
    val backtest = BacktestAPI(http)

    /** WebSocket streaming endpoints (public market data + private feeds). */
    val stream = StreamAPI(apiKey, wsUrl, http, okHttp)

    // ── Platform / agents plane ────────────────────────────────────────────────

    /**
     * Auth endpoints: login, register, MFA, password reset, session management.
     * Also includes MFA setup/confirm via [mfa].
     */
    val auth = AuthAPI(http)

    /** MFA enrollment: status, TOTP setup, confirm. */
    val mfa = MfaAPI(http)

    /**
     * Account profile, GDPR data export, CEX key removal, and credit balances.
     * For exchange key reads use [account]; for billing use [billing].
     */
    val accounts = AccountsAPI(http)

    /** Billing: subscription status, Stripe checkout / portal, pricing plans. */
    val billing = BillingAPI(http)

    /** Runner token management: mint, list, revoke `mel_run_` tokens. */
    val runner = RunnerAPI(http)

    /** Create and list agent projects. */
    val projects = ProjectsAPI(http)

    /** Pipeline run overview, traces, schedules, model prices, and server version. */
    val pipelines = PipelinesAPI(http)

    /** Human-in-the-loop approval queue: list pending, approve, reject, inspect run tool calls. */
    val hitl = HitlAPI(http)

    /** User-scoped credential storage (services, OAuth, RAG, env handles, AI models). */
    val credentials = CredentialsAPI(http)

    /** Project-scoped connector credentials (per-project service keys). */
    val connectors = ConnectorsAPI(http)

    /** Phone device control: pair, list, screen-tree, apps, active run registration. */
    val phone = PhoneAPI(http)

    /** Project team management: members, roles, invite links, pipeline visibility. */
    val team = TeamAPI(http)

    /** Pipeline templates: create, share, assign, and manage visibility. */
    val templates = TemplatesAPI(http)

    /** Assistant onboarding profile (get + set). */
    val assistant = AssistantAPI(http)

    /** Agent evaluation runs and benchmark scores. */
    val evals = EvalsAPI(http)

    /** In-app bug reports and notifications. */
    val bugs = BugsAPI(http)

    /**
     * Platform real-time events over Socket.IO at `/api/v1/events`.
     *
     * Subscribe to run updates, init-phase progress, HITL notifications,
     * and pipeline CRUD events. The connection opens lazily on the first
     * subscription or room join — constructing [Melaya] performs no events
     * I/O, so REST-only usage never opens a socket.
     *
     * Always call [MelayaEvents.close] when done, or use [AutoCloseable] with
     * try-with-resources.
     */
    val events = MelayaEvents(baseUrl, apiKey, okHttp)

    // ── Domain namespaces (primary API) ────────────────────────────────────────

    /**
     * Trading plane: market data, account, paper trading, live trading,
     * strategies, backtesting, and WebSocket streaming.
     *
     * ```kotlin
     * melaya.trading.market.ticker("binance", "BTC/USDT", "spot")
     * melaya.trading.strategies.list()
     * melaya.trading.backtest.start(…)
     * ```
     */
    val trading = TradingNamespace(
        market     = market,
        account    = account,
        sim        = sim,
        strategies = strategies,
        trade      = trade,
        backtest   = backtest,
        stream     = stream,
    )

    /**
     * Agents plane: pipeline runs, HITL approvals, assistant, phone control,
     * and evaluations.
     *
     * ```kotlin
     * melaya.agents.pipelines.recent()
     * melaya.agents.hitl.pending()
     * melaya.agents.evals.benchmarks()
     * ```
     */
    val agents = AgentsNamespace(
        pipelines = pipelines,
        hitl      = hitl,
        assistant = assistant,
        phone     = phone,
        evals     = evals,
    )

    /**
     * Platform plane: projects, credentials, connectors, billing, team,
     * templates, runner tokens, auth, MFA, accounts, bugs, and real-time events.
     *
     * ```kotlin
     * melaya.platform.projects.list()
     * melaya.platform.credentials.set("openai", value = "sk-…")
     * melaya.platform.events.onRunUpdate("run-123") { … }
     * ```
     */
    val platform = PlatformNamespace(
        projects    = projects,
        credentials = credentials,
        connectors  = connectors,
        billing     = billing,
        team        = team,
        templates   = templates,
        runner      = runner,
        auth        = auth,
        mfa         = mfa,
        accounts    = accounts,
        bugs        = bugs,
        events      = events,
    )

    // ── OkHttp client builder ──────────────────────────────────────────────────

    private fun buildOkHttpClient(): OkHttpClient {
        return OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .addInterceptor(RetryInterceptor())
            .build()
    }
}
